src/main/kotlin/graphql/kickstart/tools/MethodFieldResolver.kt

```diff
@@ -105,7 +122,18 @@ internal class MethodFieldResolver(
                     return@add value
                 }
 
-                return@add mapper.convertValue(value, typeReference)
+                return@add mapper.convertValue(value, typeReference).also {
+                    if (shouldValidate) {
+                        if (validator == null) {
+                            throw ResolverError("Unable to validate input because a Validator instance was not configured in the SchemaParserOptions.")
+                        }
```

Comment:
I figured it would make most sense to throw if a validator instance was not provided, but the `@Valid` annotation is being used. Or should we perhaps use a default validator instance?

Reply:
I think what we could do here instead is to allow providing any kind of function that gets executed after the mapper converts the value.

That way you would provide this block of code when setting up the schema parser. And every user could use whatever validation or extra logic they want.

Reply:
Hmm I see. How would that SchemaParserOption API look like? You'd need access to the reflection objects in order to scan for annotations, for example to cover these cases (this PR currently does not cover this):

```java
@Singleton
public class GetPost implements GraphQLQueryResolver {
  @Override
  public User getUser(@Size(min = 5, max = 200) @NotNull String firstName) {
    ...
  }
}
```

Reply:
I kinda like how graphql-spqr enables this through its [ResolverInterceptor API](https://github.com/leangen/graphql-spqr/blob/master/src/main/java/io/leangen/graphql/execution/ResolverInterceptor.java)

Reply:
The input to the function could be pretty much all the information we already have in the `MethodFieldResolver`. So you would have access to the target class, to the actual object as instantiated by Jackson, to the GraphQLField etc. Would that be enough to execute the validation?

The resolver interceptor API looks interesting for sure.

Reply:
Hmm I think it might get a bit awkward as-is due to how you need to map the current argument through the index variable, then I think it'd be nicer to decouple unmarshalling with validation - so collect all the unmarshalled arguments and then call provided function;

```kt
data class MethodFieldResolverExecutionContext(
  val field: FieldDefinition,
  val method: Method,
  val arguments: Array<Any>
)

class MethodFieldResolver {
    override fun createDataFetcher(): DataFetcher<*> {
        ...

        val context = MethodFieldResolverExecutionContext(...)
        methodFieldResolverInterceptor.beforeInvoke(context)
        return if (batched) {
            BatchedMethodFieldResolverDataFetcher(getSourceResolver(), this.method, args, options)
        } else { ... }.also {
          methodFieldResolverInterceptor.afterInvoke(context = context, result = it)
        }
    }
}
```

Reply:
The method field resolver interceptor can be provided via SchemaParserOptions:

```kt
SchemaParserOptions.newOptions()
  .methodFieldResolverInterceptor(object : MethodFieldResolverInterceptor {
    override fun beforeInvoke(context: MethodFieldResolverExecutionContext) {}

    override fun afterInvoke(context: MethodFieldResolverExecutionContext, result: Any) {}
  })
```

Reply:
@vojtapol what do you think about something like that?

Reply:
@vojtapol I implemented the above and force pushed, what do you think?